### ISSUE TITLE
Object properties missing since 1.61.0

### ISSUE THREAD

**Issue by webmaster128** [stephenh/ts-proto#218]
I just tried to upgrade ts-proto [from 1.60.0 to 1.61.0](https://github.com/stephenh/ts-proto/compare/v1.60.0...v1.61.0) and notices that many of my tests start failing. This is because properties of objects are missing, such that Jasmine's toEqual does not find them anymore. The tests use externally generated proto serialization with default fields omitted.

It is always about default values. I wonder if there is a different way to implement https://github.com/stephenh/ts-proto/commit/136cd28e5d2e0e708d0cc150d64d890c7dfcbd37. With `hasOwnProperty` being false, the properties become pretty much invisible (e.g. for `Object.keys(a)`/`Object.entries(a)`).

------

Test fails:

```
@cosmjs/stargate: **************************************************
@cosmjs/stargate: *                    Failures                    *
@cosmjs/stargate: **************************************************
@cosmjs/stargate: 1) AuthExtension account works for unused account
@cosmjs/stargate:   - Expected object to have properties
@cosmjs/stargate:       sequence: 0
@cosmjs/stargate:   /home/circleci/project/packages/stargate/src/queries/auth.spec.ts:25:23
@cosmjs/stargate:         assert(account);
@cosmjs/stargate:         expect(account).toEqual({
@cosmjs/stargate:                         ~
@cosmjs/stargate:           address: unused.address,
@cosmjs/stargate:           // pubKey not set
@cosmjs/stargate:   internal/process/next_tick.js:68:7
@cosmjs/stargate:   jasmine-spec-reporter: unable to open 'internal/process/next_tick.js'
@cosmjs/stargate:   Error: ENOENT: no such file or directory, open 'internal/process/next_tick.js'
@cosmjs/stargate: 2) AuthExtension account works for account with pubkey and non-zero sequence
@cosmjs/stargate:   - Expected object to have properties
@cosmjs/stargate:       accountNumber: 0
@cosmjs/stargate:   /home/circleci/project/packages/stargate/src/queries/auth.spec.ts:42:23
@cosmjs/stargate:         const pubkey = encodePubkey(validator.pubkey);
@cosmjs/stargate:         expect(account).toEqual({
@cosmjs/stargate:                         ~
@cosmjs/stargate:           address: validator.delegatorAddress,
@cosmjs/stargate:           pubKey: Any.fromPartial(pubkey),
@cosmjs/stargate:   internal/process/next_tick.js:68:7
@cosmjs/stargate:   jasmine-spec-reporter: unable to open 'internal/process/next_tick.js'
@cosmjs/stargate:   Error: ENOENT: no such file or directory, open 'internal/process/next_tick.js'
@cosmjs/stargate: 3) AuthExtension unverified account works for unused account
@cosmjs/stargate:   - Expected object to have properties
@cosmjs/stargate:       sequence: 0
@cosmjs/stargate:   /home/circleci/project/packages/stargate/src/queries/auth.spec.ts:71:25
@cosmjs/stargate:           assert(account);
@cosmjs/stargate:           expect(account).toEqual({
@cosmjs/stargate:                           ~
@cosmjs/stargate:             address: unused.address,
@cosmjs/stargate:             // pubKey not set
@cosmjs/stargate:   internal/process/next_tick.js:68:7
@cosmjs/stargate:   jasmine-spec-reporter: unable to open 'internal/process/next_tick.js'
@cosmjs/stargate:   Error: ENOENT: no such file or directory, open 'internal/process/next_tick.js'
@cosmjs/stargate: 4) AuthExtension unverified account works for account with pubkey and non-zero sequence
@cosmjs/stargate:   - Expected object to have properties
@cosmjs/stargate:       accountNumber: 0
@cosmjs/stargate:   /home/circleci/project/packages/stargate/src/queries/auth.spec.ts:88:25
@cosmjs/stargate:           const pubkey = encodePubkey(validator.pubkey);
@cosmjs/stargate:           expect(account).toEqual({
@cosmjs/stargate:                           ~
@cosmjs/stargate:             address: validator.delegatorAddress,
@cosmjs/stargate:             pubKey: Any.fromPartial(pubkey),
@cosmjs/stargate:   internal/process/next_tick.js:68:7
@cosmjs/stargate:   jasmine-spec-reporter: unable to open 'internal/process/next_tick.js'
@cosmjs/stargate:   Error: ENOENT: no such file or directory, open 'internal/process/next_tick.js'
@cosmjs/stargate: 5) IbcExtension unverified connection works
@cosmjs/stargate:   - Expected object to have properties
@cosmjs/stargate:       delayPeriod: 0
@cosmjs/stargate:   /home/circleci/project/packages/stargate/src/queries/ibc.spec.ts:196:37
@cosmjs/stargate:           const response = await client.ibc.unverified.connection(ibcTest.connectionId);
@cosmjs/stargate:           expect(response.connection).toEqual(ibcTest.connection);
@cosmjs/stargate:                                       ~
@cosmjs/stargate:           expect(response.proofHeight).toBeDefined();
@cosmjs/stargate:           expect(response.proofHeight).not.toBeNull();
@cosmjs/stargate:   internal/process/next_tick.js:68:7
@cosmjs/stargate:   jasmine-spec-reporter: unable to open 'internal/process/next_tick.js'
@cosmjs/stargate:   Error: ENOENT: no such file or directory, open 'internal/process/next_tick.js'
```

**Comment by stephenh** [comment]
Shoot. Yeah, that's unfortunate re `Object.keys` and `Object.entries`, full disclosure I mis-remembered what the semantics of those were, and assumed they would include prototype's keys.

I'm pretty open to ideas here. I personally liked/was fine with the non-prototype approach, but the killer was per that commit message, protobuf's own `oneofIndex` having different semantics when set vs. not-set, and that only being observable by having the default values "somewhere else".

Historically this wasn't a problem when ts-proto used pbjs-generated code for its internal impls, but I'd been wanting/meaning to switch over to ts-proto-generated code "just because" / achieve the goal of being self-hosting, and this was the one annoying little thing that was tripping things up.

...I suppose it could be yet-another-options flag, and defaulted to not-`Object.create` b/c frankly I don't think anyone should really use it unless you very specifically need to differentiate this has-been-set case...

**Comment by webmaster128** [comment]
Could you elaborate on the problem the commit is trying to solve? Maybe we find a better way to address it. I guess it is about

```ts
export function isWithinOneOf(field: FieldDescriptorProto): boolean {
  return field.hasOwnProperty('oneofIndex');
}
```

where `FieldDescriptorProto` is a class from `build/pbjs.d.ts`/`build/pbjs.js`, which is implemented as a function with properties being set like this:

```js
                function FileDescriptorProto(properties) {
                    this.dependency = [];
                    this.publicDependency = [];
                    this.weakDependency = [];
                    this.messageType = [];
                    this.enumType = [];
                    this.service = [];
                    this.extension = [];
                    if (properties)
                        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                            if (properties[keys[i]] != null)
                                this[keys[i]] = properties[keys[i]];
                }
```

But [given that](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/field#oneofindex)

> The first type has index 1; zero means the type is not in the list. 

why not just write the check as follows?

```ts
export function isWithinOneOf(field: FieldDescriptorProto): boolean {
  return !!field.oneofIndex;
}
```

**Comment by stephenh** [comment]
> The first type has index 1; zero means the type is not in the list.

I'd read that too, but that was not the observed behavior. I'd written `field.oneofIndex > 0`, essentially the same thing, but for this `first` field:

https://github.com/stephenh/ts-proto/blob/master/integration/simple/simple.proto#L72

The `oneofIndex` was coming back as `0`.

I'd love to be missing something, but I'm not sure why those docs don't match what was showing up in the message.

**Comment by webmaster128** [comment]
I see. The docs I linked above are C#. No idea if the it can be applied to JS as well.

Looking at `protos/google/protobuf/descriptor.proto`, we have

```
  // If set, gives the index of a oneof in the containing type's oneof_decl
  // list.  This field is a member of that oneof.
  optional int32 oneof_index = 9;
```

so this is an optional int (which is different than the C# docs). So `public oneofIndex: number;` in `build/pbjs.d.ts` seems wrong. No idea how optional numbers are supposed to be implemented in pbjs. But what about this?

```
export function isWithinOneOf(field: FieldDescriptorProto): boolean {
  return typeof field.oneofIndex === "number";
}
```

**Comment by stephenh** [comment]
> public oneofIndex: number

Thought about this, and my initial reaction was "well in proto3 you can't really have primitives as `number | undefined` because of the default-values-aren't-serialized-on-the-wire-so-you-can't-do-presence checks".

But maybe what I've been missing is that this `FieldDescriptorProto` is actually a proto2 file, where optional was still a thing, such that, I think you're right, this field's type should come out as `oneofIndex: number | undefined`, with no default value applied (whether via the prototype or object spread), which would allow detecting presence with a simple `oneofIndex !== undefined` check.

There's an existing issue to support proto2 optional fields:

https://github.com/stephenh/ts-proto/issues/139

So I'm hoping that if we implemented that, and got this `oneofIndex` over to a `number | undefined`, we should be able to go back to object spread / no `Object.create`. Seems right?

I can poke around at this...it may be as simple updating:

https://github.com/stephenh/ts-proto/blob/master/src/main.ts#L421

To look at the proto2 optional label, and then probably a few updates that fall out from that initial change.

It might be a few days before I get to it, if you have time sooner than that, do you want to take a look? Either's good with me.

**Comment by stephenh** [comment]
Okay, v1.66.0 reverted back to object spread.

I didn't technically fix #139 yet, however since `ts-proto-dependencies` is a published npm artifact, ts-proto can keep using its `Object.create`-using messages for parsing the `protoc` plugin messages, until for whatever future reason we need to regenerate/republish those messages.

**Comment by webmaster128** [comment]
Works like a charm, thank you!
